### ISSUE TITLE
_clear(desc) Not Clearing Whole Caption Text Box If Title To To Long

### ISSUE THREAD

**Issue by Jamadoo** [wkaisertexas/tiktok-uploader#131]
This is a issue relating to @callum-fortune 's new Pull-Request

On Lines
```python
    # desc populates with filename before clearing
    WebDriverWait(driver, config['explicit_wait']).until(lambda driver: desc.text != '')

    _clear(desc)

    WebDriverWait(driver, config['explicit_wait']).until(lambda driver: desc.text == '')
```

If the title of the file is longer then the half of the text box, only the text before the half of the text box is removed. Considor sending "Keys.End" before calling clear

```python
    # desc populates with filename before clearing
    WebDriverWait(driver, config['explicit_wait']).until(lambda driver: desc.text != '')

    desc.send_keys(Keys.END)
    _clear(desc)

    WebDriverWait(driver, config['explicit_wait']).until(lambda driver: desc.text == '')
```

Works great in my experience.

I also noticed a log is send in the console of the browser ({isSameSearchValue: false}) each time the dialog pops up showcasing the available hashtags and mentions. We could use a wait until this log gets written instead of
```python
if word[0] == "#":
                desc.send_keys(word)
                desc.send_keys(' ' + Keys.BACKSPACE)
                time.sleep(config['implicit_wait']) # Here we would add the wait until the log gets written, aka the menu gets opened. Which will speed up the hashtag adding progress a ton, because the menu pops up between 0.5-2 seconds in my experience. Thus, here we wont wait any  unnecessary time 
                desc.send_keys(Keys.ENTER)
```

**Comment by Jamadoo** [comment]
@callum-fortune It was something along the lines "you forgot your cocaine at the Whitehouse, what's your next move_" (it only sounds weird with no context). When I ran the script the textbox gets clicked, but in the middle of the textbox so the cursor is in the middle of the textbox. Then clear gets called which only presses backspace. Thus leaving the text after the cursor in tacit and putting the script in a infinite loop waiting for the textbox to be empty. 

The script is a bit unpredictable some times in my experience, so it could be one of those instances. Might as well add the keys.end as a safety procedure

**Comment by Jamadoo** [comment]
I got a suggestion from gpt how to monitor the logs and wait for a print. I'll see what I can implement later today.

**Comment by callum-fortune** [comment]
@Jamadoo I must confess, I tried one or two ideas from chatGPT but none of them pulled through. I also think there could be cross-browser discrepancies between how the logs are read which could cause issues in the future. With that said though, go for it, I'd be interested to see how it works in the end.

**Comment by Jamadoo** [comment]
yea, the suggestion is focused on the chrome driver. So if anything, we can enable the log wait on chorme or else the method you mentioned. Then we just say "we recommend using Chrome driver". I can't really see why you would want to use another browser tbh

**Comment by Jamadoo** [comment]
I found a potential way to check the logs in chromium based browsers. We could also just remove support to Firefox 😂😭

**Comment by callum-fortune** [comment]
@Jamadoo I think the recommendation is already to use Chrome based browsers so yeh it would be dumb to use anything else, I guess I'm trying to cover all bases really. I'm looking forward to seeing a working example 😃. I've dropped you a follow for the help.

**Comment by Jamadoo** [comment]
Im looking into this wait for a log thing. I got it working to get all browser logs and wait for out specific log. Then I realized that this log is send through JavaScript. And according to gpt and some of my own research you need to set up fucking websockets and shit to capture this. I aint doing allat. Ill be looking at other ways to detect when the menu pops up.

**Comment by callum-fortune** [comment]
@Jamadoo Yeh thats the thing, it's so extra. But the PR that is currently open (https://github.com/wkaisertexas/tiktok-uploader/pull/128) I found a way to detect the menu popup and it works fine so currently there is no need to look at the logs anyway. Maybe in the future as the UI changes more that would be necessary... who knows. Hopefully @wkaisertexas can take a look at the new PR and we can get it merged. Initially I was planning to use this project for a personal side-project but I'm quite enjoying looking into and fixing various issues tbh.

For reference, the change I made to detect the popup is in `config.toml` and now references the popup using the following selector:

`mention_box = "//div[contains(@class, 'mention-list-popover')]"`

Despite the name, `mention-list-popover` is used for both hashtags and account mentions

**Comment by Jamadoo** [comment]
Oh, thats quite literally what ive been looking for.... I went into the site's javascript and found they are setting the css data of a element in the document, but could not for the life of me find it. I believe they toggle it's visibility (display type). 

In the PR, i see you create the EC.presence_of_element_located, but never wait for it (Note, i code in c# not python), is this correct or should there be a WebDriverWait?
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/9604a987-5dc6-48c9-9a04-b3a36dac1473)

**Comment by Jamadoo** [comment]
Also, @callum-fortune I found on slower internet connecitons ['selectors']['upload']['upload_video'] is not immediately accessible inside the site and cant be send keys to on `_set_video`. I found just adding another wait before sending the keys fixed this

```python
            _change_to_upload_iframe(driver)

            # Wait For Input File
            driverWait = WebDriverWait(driver, config['explicit_wait'])
            upload_boxWait = EC.presence_of_element_located(
                (By.XPATH, config['selectors']['upload']['upload_video'])
                )
            driverWait.until(upload_boxWait)
```

**Comment by Jamadoo** [comment]
@callum-fortune Just tested the code with 

```python
               driverWait.until(EC.presence_of_element_located(
                    (By.XPATH, config['selectors']['upload']['mention_box'])
                ))
                time.sleep(0.5) # for safty
```

Works better then i ever expected

**Comment by callum-fortune** [comment]
@Jamadoo As per your first comment, You're totally right and I completely overlooked putting the element selector in a webDriverWait, I'm just sorting that now. As for the explicit wait for slow internet connections I'll add that to the PR too

**Comment by Jamadoo** [comment]
@callum-fortune  The `time.sleep(0.5) # For Safety` is not even needed. I tested it with my shitty ass 500 kb/s mobile wifi, and it still timed the `keys.enter` to the millisecond. Thats all we needed

**Comment by callum-fortune** [comment]
@Jamadoo That's great I just tried it too and it's super fast. There is an issue with the @ mentions though where if I remove the various time.sleep(1) it doesn't load the correct account in time and selects an incorrect one. Either we keep the time.sleeps in there or figure something out...

**Comment by Jamadoo** [comment]
Is the mention_box element a flex object? If it is, we can look at the amount of elements/children of the flex object and just wait untill theres more then two. I cant for the life of me find where the hell this mention_box element is, so could you also show me your inspect element window of where this mf is 😭

**Comment by Jamadoo** [comment]
@callum-fortune Thanks, ill quickly look into its children

**Comment by Jamadoo** [comment]
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/9760ffb2-dfe8-4c5d-ad7a-98a6ed411310)
🤗 looks like we got a div with children

**Comment by Jamadoo** [comment]
This div only appear AFTER all the names are loaded. So we just need to wait untill this div appear inside the Popover element

**Comment by callum-fortune** [comment]
@Jamadoo Which div would that be? When I watch the script running, I see that several users come up but not one that is an exact match of the mention, it takes a few secs for the exact match to load in, meaning it ends up selecting whichever account is at the top of the list at that moment

**Comment by Jamadoo** [comment]
@callum-fortune  On my side, when i type in any known taken username, tiktok gives me a loading animation. In this loading animation the div (with no attributes) is not displayed yet. Only when the usernames are loaded the div gets shown, and the exact match is at the top. Ill make a quick video with a username i have not searched yet

**Comment by Jamadoo** [comment]
@callum-fortune Look here. My first attampt, you can see i search a random tiktok username and the axact match comes up right after the loading completes. Then the console gets paused, and we can see theres a div with all the users in. Then my 3rd attempt (ignore me struggling the the second attempt) while the usernames are loading the div is not created yet.
https://gofile.io/d/kfez70

**Comment by callum-fortune** [comment]
@Jamadoo I think we can just create a wait for the element in the popover that contains the username we are searching for since ti is displayed as a heading or whatever. I think this was done in the original implementation for this repo but when I refactored one of the big functions this would have been lost

**Comment by Jamadoo** [comment]
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/c8e2f7f3-2fd1-497b-9d89-02edd1fc8380)
We could, here is where the username is stored. I do see this as unnecessary tho? Or is it cause my cache is already loaded for the usernames to match first time?

**Comment by callum-fortune** [comment]
@Jamadoo Yeh I think it is a caching issue because 9/10 times it works perfectly for me as well but it is probably worth making the change. Also, the code would need to reference the user-id field not the name which creates another issue... check the contents of the user-id span...

**Comment by Jamadoo** [comment]
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/bbd74421-4c69-4fd7-8f9f-c779e7ba49e4)
this is how the userId looks, we can just search for the user-id class, then select the first child. Then repeat this untill the first one is out user. And have a timeout of 5 seconds, we select the first one (we assume the user gave a invalid username)

**Comment by Jamadoo** [comment]
Got code working to wait for the div and then wait for 2 children to appear
```python
# Wait For Popup
                Popup = driverWait.until(EC.presence_of_element_located(
                    (By.XPATH, config['selectors']['upload']['mention_box'])
                ))
                # Wait For Div With Users
                Div = driverWait.until(
                    lambda driver: Popup.find_element(By.XPATH, "./div[not(@*)]")
                )
                # Wait until it has at least two children
                driverWait.until(
                    lambda driver: len(Div.find_elements(By.XPATH, "./*")) >= 2
                )
                logger.debug(green('- Loaded Users'))
```

**Comment by callum-fortune** [comment]
@Jamadoo What I interpreted from your message above reading 'we can just search for the user-id class, then select the first child' seems like the better idea, I'm just testing that and then we can compare

**Comment by Jamadoo** [comment]
@callum-fortune  got my code mostly working for my idea. Just some bugs to get killed

**Comment by Jamadoo** [comment]
@callum-fortune  Chatgpt really saving my ass rn ngl

**Comment by callum-fortune** [comment]
@Jamadoo With your implementation, what will happen if the search is for say "bbc" and someone has the username "abbca". Will your code only match it exactly and return the correct element for @bbc or will it return others that contain the string and perhaps return @abbca

**Comment by Jamadoo** [comment]
@callum-fortune In my code im looking for a exact match (not case sens). If non are found in 5 secs after the usernames are loaded. I select the top username

**Comment by Jamadoo** [comment]
I got some working code, testing everything rn

**Comment by callum-fortune** [comment]
@Jamadoo Yeh that sounds much better than my implementation tbh

**Comment by Jamadoo** [comment]
@callum-fortune I got some working code
```python
 elif word[0] == "@":
                logger.debug(green('- Adding Mention: ' + word))
                desc.send_keys(word)
                desc.send_keys(' ' + Keys.BACKSPACE)
                # Wait For Popup
                Popup = driverWait.until(EC.presence_of_element_located(
                    (By.XPATH, config['selectors']['upload']['mention_box'])
                ))
                # Wait For Div With Users
                Div = driverWait.until(
                    lambda driver: Popup.find_element(By.XPATH, "./div[not(@*)]")
                )
                # Wait until it has at least two children
                driverWait.until(
                    lambda driver: len(Div.find_elements(By.XPATH, "./*")) >= 2
                )
                ## Wait For Correct User To Appear
                # Variable indicating whether the loop should continue
                found = False
                waiting_interval = 0.5
                timeout = 5
                start_time = time.time()
                # Loop until the desired condition is met
                while not found and (time.time() - start_time < timeout):
                    # Get all children with 'user-id' class
                    user_id_elements = Div.find_elements(By.CLASS_NAME, "user-id")
                    
                    if user_id_elements:
                        # If elements are found, check the first one
                        first_user_id = user_id_elements[0]

                        # Get the inner text
                        first_user_id_text = first_user_id.text
                        username = first_user_id_text.split(" · ")[0]
                        print("Userrname: " + username + "\nSeaerch: " + word[1:])
                        
                        # Compare with the target word
                        if username.lower == word[1:].lower:
                            found = True
                            print("Matching User found")
                        else:
                            # If not matching, wait and repeat
                            print(f"No match. Waiting for {waiting_interval} seconds...")
                            time.sleep(waiting_interval)  # Pause for the specified interval
                    else:
                        # If no 'user-id' elements are found, wait and try again
                        print(f"No elements found. Waiting for {waiting_interval} seconds...")
                        time.sleep(waiting_interval)  # Pause before checking again
                logger.debug(green('- Selecting First User'))
                time.sleep(10)
                desc.send_keys(Keys.ENTER)
```

**Comment by callum-fortune** [comment]
@Jamadoo Cool I gave it a try, it did work however I noticed something. I passed @bbc into the description and it kept saying no match even though it had come up. In the end it selected @bbc but only because it was the first option and it timed out...

**Comment by Jamadoo** [comment]
@callum-fortune jip, it was because of my line
if username.lower() == word[1:].lower():

I forgot the () after lower. Im testing again rn.

**Comment by Jamadoo** [comment]
works perfectly now

**Comment by callum-fortune** [comment]
@Jamadoo Sounds good I'm running it now too

**Comment by callum-fortune** [comment]
@Jamadoo I have bad news, the problem seems to have persisted. I passed in @bbc and it uploaded with something totally wrong

**Comment by Jamadoo** [comment]
strange, i did the same with bbc and the correct use was added

**Comment by callum-fortune** [comment]
This is the thing, it seems dependent on many factors, environment, internet speed, cache etc

**Comment by Jamadoo** [comment]
all those waits and checks i do believe we have covered most of that? Im checking the bbc example again with some other tests and ill update if i can replicate the issue

**Comment by Jamadoo** [comment]
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/e2d77e2d-8c5b-4a9d-af98-247d4d27e40e)
bbc was selected for me....

**Comment by Jamadoo** [comment]
are you sure your if statements looks like:

```python
if username.lower() == word[1:].lower():
```

**Comment by Jamadoo** [comment]
![image](https://github.com/wkaisertexas/tiktok-uploader/assets/79752368/a058aaa1-760e-497a-b843-5e09df4f22a1)
Was uploaded with Mention as well

**Comment by Jamadoo** [comment]
ive tested this code with my 50 mpb/s fiber and my 500 kb/s mobile data. Seems to work perfectly on both....

**Comment by callum-fortune** [comment]
@Jamadoo I missed two parentheses, I added the one that you said but not one a few characters to the left...

Testing now

**Comment by Jamadoo** [comment]
> @Jamadoo I have bad news, the problem seems to have persisted. I passed in @bbc and it uploaded with something totally wrong
> 
> You can see the username that it ended up uploading:
> 
> ![image](https://private-user-images.githubusercontent.com/63158857/324274857-3e6920b1-a752-45f0-b07a-d68ddcf2f86c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3Mjc1NTgsIm5iZiI6MTcxMzcyNzI1OCwicGF0aCI6Ii82MzE1ODg1Ny8zMjQyNzQ4NTctM2U2OTIwYjEtYTc1Mi00NWYwLWIwN2EtZDY4ZGRjZjJmODZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIxVDE5MjA1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkZjgwNGY2YmFlZmZhOTgxYjVmNjMzZmViNmQzY2ZiMjc0NjRiOTI3OTE3ODlhMThmZmRiOWJiMTM2YWQ1YjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ad8munnQZKVJog2nMciayuyAIiF-Scx-12zmIekPg40)

what seems strange is the first user in the list of user which tiktok gives is anthony... How the hell does tiktok get that>

**Comment by callum-fortune** [comment]
@Jamadoo I know right I think his username has TikTok in it and they match both when searching

**Comment by Jamadoo** [comment]
@callum-fortune but thats suppose to be a issue on tiktok's side? What we can do is, we loop through all the children of the div. Check the username. If the user-ids match, we click the element. And after 5 seconds, we just press enter to select the first element

**Comment by callum-fortune** [comment]
@Jamadoo This time I got @cooplive returned lol. Not sure whats going on, I'm debugging it now

**Comment by Jamadoo** [comment]
@callum-fortune are you running on normal mod (not headless)? What does tiktok give in the list of usernames when the bot enters bbc?

**Comment by callum-fortune** [comment]
@Jamadoo Yeh I'm not running headless at the moment so I can see whats happening. Initially it was giving random users that I follow so i added a time.sleep between sending the space and backspace. Then it started returning bbc accounts but still selected @bbcnews instead of @bbc. Weird...

**Comment by Jamadoo** [comment]
try increasing the timeout to like 30 secs and remove your sleep. See if the users get updated

**Comment by callum-fortune** [comment]
@Jamadoo I was literally just coming to say that, you were only checking the first user

**Comment by Jamadoo** [comment]
i think i got it wokring. Still debugging @callum-fortune

**Comment by callum-fortune** [comment]
@Jamadoo I think the reason you couldn't reproduce is because the top recommendations are people that you follow, I assume you are using a tiktok account that follows nobody but I follow many people.

A also have a solution that appears to work, will be interesting to see how they compare 👀

**Comment by Jamadoo** [comment]
@callum-fortune yes, exactly. Im using a account with no followers. I could quickly test with my main

What your solution?

**Comment by callum-fortune** [comment]
@Jamadoo My solution counts the iterations in the loop and when it finds the correct match, it sends the down key that many times to navigate down the dropdown and select the correct user

**Comment by Jamadoo** [comment]
@callum-fortune Im simply getting all the user-id classes and looping through them.
Heres my code
```python
            elif word[0] == "@":
                logger.debug(green('- Adding Mention: ' + word))
                desc.send_keys(word)
                desc.send_keys(' ' + Keys.BACKSPACE)
                # Wait For Popup
                Popup = driverWait.until(EC.presence_of_element_located(
                    (By.XPATH, config['selectors']['upload']['mention_box'])
                ))
                # Wait For Div With Users
                Div = driverWait.until(
                    lambda driver: Popup.find_element(By.XPATH, "./div[not(@*)]")
                )
                # Wait until it has at least two children
                driverWait.until(
                    lambda driver: len(Div.find_elements(By.XPATH, "./*")) >= 2
                )
                ## Wait For Correct User To Appear
                # Variable indicating whether the loop should continue
                found = False
                waiting_interval = 0.5
                timeout = 5
                start_time = time.time()
                # Loop until the desired condition is met
                while not found and (time.time() - start_time < timeout):
                    # Get all children with 'user-id' class
                    user_id_elements = Div.find_elements(By.CLASS_NAME, "user-id")
                    
                    for user_id_element in user_id_elements:
                        if user_id_element and user_id_element.is_enabled:
                            # If elements are found, check the first one
                            first_user_id = user_id_elements[0]

                            # Get the inner text
                            first_user_id_text = first_user_id.text
                            username = first_user_id_text.split(" · ")[0]
                            
                            # Compare with the target word
                            if username.lower() == word[1:].lower():
                                found = True
                                user_id_element.click()
                                print("Matching User found : Clicking User")
                                break
                    if not found:
                        # If not matching, wait and repeat
                        print(f"No match. Waiting for {waiting_interval} seconds...")
                        time.sleep(waiting_interval)  # Pause for the specified interval
                if not found:
                    logger.debug(green('- Selecting First User'))
                    desc.send_keys(Keys.ENTER)
```

**Comment by Jamadoo** [comment]
> @Jamadoo My solution counts the iterations in the loop and when it finds the correct match, it sends the down key that many times to navigate down the dropdown and select the correct user

so you loop through the user-id and count how many down is the needed user and the press down arrow that many times? Im geussing you dont know you can just click that element 😂

**Comment by Jamadoo** [comment]
@callum-fortune  dont worry, its also 10pm for me

**Comment by Jamadoo** [comment]
im starting to loose all my focus as well 😂

**Comment by Jamadoo** [comment]
last test run, then i need to do some work for my classes tmr. And also summarize a whole half a chapter of science for tmr 😂

**Comment by callum-fortune** [comment]
@Jamadoo Where are you actually located? its 9pm here. I'm just implementing your code on my side and I'll get it all cleaned up and pushed to that PR

**Comment by Jamadoo** [comment]
@callum-fortune Im from South Africa. and u?

Great, so im guessing the script works for situation?

**Comment by callum-fortune** [comment]
@Jamadoo  I'm in the UK. Yeh it pretty much works now just a few touches, I'll make sure to co-author you in the PR

**Comment by Jamadoo** [comment]
i just tested the script on my main, works like excepted as well. I did notice my US account and SA account's UI was different. That could also be a factor for why random users popup for you. @callum-fortune

**Comment by Jamadoo** [comment]
> @Jamadoo I'm in the UK. Yeh it pretty much works now just a few touches, I'll make sure to co-author you in the PR

alright, sounds good. Thanks for your help. Ima leave the rest to you

**Comment by callum-fortune** [comment]
@Jamadoo Cheers, have a good one!

**Comment by Jamadoo** [comment]
@callum-fortune Thanks, you to 🍻
